File: src/main/java/de/maxhenkel/pipez/ItemFilter.java
package de.maxhenkel.pipez;

import com.mojang.serialization.Codec;
import de.maxhenkel.corelib.tag.SingleElementTag;
import de.maxhenkel.corelib.tag.Tag;
import de.maxhenkel.corelib.tag.TagUtils;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.world.item.Item;
import org.jetbrains.annotations.Nullable;

import java.util.UUID;

public class ItemFilter extends Filter<ItemFilter, Item> {

    private static final TagConverter<Item> TAG_CONVERTER = (single, location) -> {
        if (single) {
            return new SingleElementTag<>(location, BuiltInRegistries.ITEM.get(location));
        } else {
            return TagUtils.getItemTag(location);
        }
    };
    public static final Codec<Tag<Item>> TAG_CODEC = tagCodec(TAG_CONVERTER);
    public static final StreamCodec<RegistryFriendlyByteBuf, Tag<Item>> STREAM_TAG_CODEC = tagStreamCodec(TAG_CONVERTER);

    public static final Codec<ItemFilter> CODEC = codec(ItemFilter.class, TAG_CODEC);
    public static final StreamCodec<RegistryFriendlyByteBuf, ItemFilter> STREAM_CODEC = streamCodec(ItemFilter.class, STREAM_TAG_CODEC);

    public ItemFilter(UUID id, @Nullable Tag<Item> tag, @Nullable CompoundTag metadata, boolean exactMetadata, @Nullable DirectionalPosition destination, boolean invert) {
        super(id, tag, metadata, exactMetadata, destination, invert);
    }

    @Override
    public Codec<ItemFilter> getCodec() {
        return CODEC;
    }

    @Override
    public StreamCodec<RegistryFriendlyByteBuf, ItemFilter> getStreamCodec() {
        return STREAM_CODEC;
    }

    public ItemFilter() {
        this(UUID.randomUUID(), null, null, false, null, false);
    }

}

File: src/main/java/de/maxhenkel/pipez/ItemFilter.kt
package de.maxhenkel.pipez

import de.maxhenkel.corelib.tag.SingleElementTag
import de.maxhenkel.corelib.tag.TagUtils
import de.maxhenkel.pipez.types.DirectionalPosition
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.Tag
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.Item
import net.minecraftforge.registries.ForgeRegistries
import java.util.*

class ItemFilter : Filter<Item>() {
    override fun serializeNBT(): CompoundTag {
        val compound = CompoundTag()
        tag?.also {
            if (it is SingleElementTag) {
                ForgeRegistries.ITEMS.getKey(
                    (it as SingleElementTag<Item>).element
                )?.also { key ->
                    compound.putString("Item", key.toString())
                }
            } else {
                compound.putString("Tag", it.name.toString())
            }
        }
        metadata?.also {
            compound.put("Metadata", it)
        }
        if (exactMetadata) {
            compound.putBoolean("ExactMetadata", true)
        }
        destination?.also {
            compound.put("Destination", it.serializeNBT())
        }
        if (invert) {
            compound.putBoolean("Invert", true)
        }
        compound.putUUID("ID", id)

        return compound
    }

    override fun deserializeNBT(compound: CompoundTag) {
        tag = null
        if (compound.contains("Item", Tag.TAG_STRING.toInt())) {
            val item = ForgeRegistries.ITEMS.getValue(
                ResourceLocation(compound.getString("Item"))
            )
            if (item != null) {
                tag = SingleElementTag(item.registryName, item)
            }
        }
        if (compound.contains("Tag", Tag.TAG_STRING.toInt())) {
            tag = TagUtils.getItemTag(ResourceLocation(compound.getString("Tag")))
        }

        metadata = if (compound.contains("Metadata", Tag.TAG_COMPOUND.toInt())) {
            compound.getCompound("Metadata")
        } else {
            null
        }

        destination = if (compound.contains("Destination", Tag.TAG_COMPOUND.toInt())) {
            DirectionalPosition.fromNBT(compound.getCompound("Destination"))
        } else {
            null
        }

        invert = if (compound.contains("Invert", Tag.TAG_BYTE.toInt())) {
            compound.getBoolean("Invert")
        } else {
            false
        }

        id = if (compound.contains("ID", Tag.TAG_INT_ARRAY.toInt())) {
            compound.getUUID("ID")
        } else {
            UUID.randomUUID()
        }

    }
}